src/Access/ContextAccess.cpp

```diff
@@ -146,17 +146,19 @@ ContextAccess::ContextAccess(const AccessControl & access_control_, const Params
     : access_control(&access_control_)
     , params(params_)
 {
-    std::lock_guard lock{mutex};
+    setUser(access_control->read<User>(*params.user_id));
+}
 
-    subscription_for_user_change = access_control->subscribeForChanges(
-        *params.user_id, [this](const UUID &, const AccessEntityPtr & entity)
-    {
-        UserPtr changed_user = entity ? typeid_cast<UserPtr>(entity) : nullptr;
-        std::lock_guard lock2{mutex};
-        setUser(changed_user);
-    });
 
-    setUser(access_control->read<User>(*params.user_id));
+void ContextAccess::subscribeForChanges(const std::shared_ptr<ContextAccess> &res)
+{
+     subscription_for_user_change = access_control->subscribeForChanges(
+         *params.user_id, [res](const UUID &, const AccessEntityPtr & entity)
```

Comment:
This is not correct. You pass `res` (which is a shared pointer to `ContextAccess`) to the lambda function and the lambda will keep it. On the other hand `ContextAccess` will own the lambda via the supscription because `ContextAccess` has `subscription_for_user_change` in its member variables. So `ContextAccess` will own the lambda and the lambda will own the `ContextAccess`, and that cyclic ownerships will reason that the `ContextAccess` will never be destroyed. That's not right.

In order to fix that we can use weak pointers, I mean we can pass a weak pointer to the lambda if you write

```
[weak_ptr = weak_from_this()]
```

instead of `[res]`, and then you can check the weak pointer inside the lambda

```
auto ptr = weak_ptr.lock();
if (!ptr)
    return;
...
ptr->setUser(changed_user);
```

And it's better to inherit `ContextAccess` from `std::enable_shared_from_this<ContextAccess>` - this would allow not to pass a shared pointer to `ContextAccess` as a parameter to `ContextAccess::subscribeForChanges()`.

Reply:
OK ， I will work on this.